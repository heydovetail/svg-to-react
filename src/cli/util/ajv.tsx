import Ajv = require("ajv");
import chalk from "chalk";
import * as fs from "fs";

export function formatErrorMessage(path: string, validate: Ajv.ValidateFunction): string {
  return `Invalid config ${path}, ${validate
    .errors!.map(e => (e.message !== undefined ? chalk`> {red ${e.message}}` : ""))
    .join("\n")}`;
}

export function readFileOrThrow(path: string, validate: Ajv.ValidateFunction): object {
  const config = JSON.parse(fs.readFileSync(path, "utf8"));
  if (validate(config)) {
    return config;
  } else {
    throw new Error(formatErrorMessage(path, validate));
  }
}
