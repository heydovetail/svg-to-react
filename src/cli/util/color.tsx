export function normalizeHex(hex: string): string {
  const compressableHexMatch = hex.match(/^#(([a-f0-9])\2){3}$/i);
  if (compressableHexMatch !== null) {
    const [str] = compressableHexMatch;
    return `#${str[1]}${str[3]}${str[5]}`.toLowerCase();
  }

  const hexMatch = hex.match(/^#([a-f0-9]{3}){1,2}$/i);
  if (hexMatch !== null) {
    const [str] = hexMatch;
    return str.toLowerCase();
  }

  return hex;
}
