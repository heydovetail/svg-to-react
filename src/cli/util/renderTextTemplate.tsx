/**
 * Render a text template that supports replacing variables with dynamic values.
 * @param template Template that uses braces to interpolate variables. Variables
 * are demarcated by `{` and `}`. Backslash can be used to escape characters.
 * @param context An object mapping variable names to values.
 */
export function renderTextTemplate(template: string, context: { [key: string]: string | undefined }): string {
  const enum ParseContext {
    LITERAL,
    VARIABLE
  }

  const enum Char {
    ESCAPE = "\\",
    VARIABLE_START = "{",
    VARIABLE_END = "}"
  }

  let parseContext = ParseContext.LITERAL;
  let escape = false;
  let result = "";
  let variableName = "";

  for (const char of template) {
    switch (parseContext) {
      case ParseContext.LITERAL: {
        if (escape) {
          result += char;
          escape = false;
        } else if (Char.ESCAPE === char) {
          escape = true;
        } else if (Char.VARIABLE_START === char) {
          parseContext = ParseContext.VARIABLE;
        } else {
          result += char;
        }
        break;
      }
      case ParseContext.VARIABLE: {
        if (escape) {
          variableName += char;
          escape = false;
        } else if (Char.ESCAPE === char) {
          escape = true;
        } else if (Char.VARIABLE_END === char) {
          const value = context[variableName];
          if (value !== undefined) {
            result += value;
          }
          variableName = "";
          parseContext = ParseContext.LITERAL;
        } else {
          variableName += char;
        }
        break;
      }
      default: {
        if (process.env.NODE_ENV !== "production") {
          const exhausted: never = parseContext;
          exhausted;
        }
      }
    }
  }

  if (variableName !== "") {
    result += Char.VARIABLE_START;
    result += variableName;
  }

  return result;
}
