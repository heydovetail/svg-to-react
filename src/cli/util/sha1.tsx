import * as crypto from "crypto";

export function sha1(data: string) {
  const generator = crypto.createHash("sha1");
  generator.update(data);
  return generator.digest("hex");
}
