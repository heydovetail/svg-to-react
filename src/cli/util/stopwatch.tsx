export async function stopwatch(fn: () => Promise<void>): Promise<number> {
  const startTime = new Date();
  await fn();
  const endTime = new Date();
  return endTime.getTime() - startTime.getTime();
}
