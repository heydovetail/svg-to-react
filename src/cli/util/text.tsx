export function firstUpperCase(text: string) {
  return `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
}

export function hyphenToCamelCase(text: string) {
  return text.replace(/[-_](.)/g, (_, chr) => chr.toUpperCase());
}

export function fileNameToComponentName(fileName: string) {
  return firstUpperCase(hyphenToCamelCase(fileName));
}
