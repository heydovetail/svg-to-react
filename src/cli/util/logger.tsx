import chalk from "chalk";
import { stopwatch } from "./stopwatch";

export interface Logger {
  write: (text: string) => void;
}

export async function logTask(name: string, task: () => Promise<void>, logger: Logger) {
  logger.write(chalk`{yellow >} ${name}`);
  const duration = await stopwatch(task);
  logger.write(chalk`\r{green âœ” ${name}} {grey (${`${duration}`}ms)}\n`);
}
