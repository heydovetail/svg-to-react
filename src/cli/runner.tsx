import * as fs from "fs";
import { createPlan } from "./planner";
import { transform } from "./transform";
import { Logger, logTask } from "./util/logger";

export interface TextFileWriter {
  (path: string, content: string): void;
}

export async function runner(
  rootPath: string,
  writeFile: TextFileWriter = fs.writeFileSync,
  logger: Logger = process.stdout
): Promise<void> {
  const plan = await createPlan(rootPath);

  logger.write(`Generating React components for ${plan.items.length} SVGs:\n`);

  for (const { svgPathRelative, modulePath, config, svgPath } of plan.items) {
    await logTask(
      svgPathRelative,
      async () => {
        const esModule = await transform(fs.readFileSync(svgPath, "utf8"), config);
        writeFile(modulePath, esModule);
      },
      logger
    );
  }
}
