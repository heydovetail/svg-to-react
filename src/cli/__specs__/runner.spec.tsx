import "jest";
import * as path from "path";
import * as runner from "../runner";
import * as stopwatch from "../util/stopwatch";

class TestTextFileWriter {
  public files = new Map<string, string>();

  public constructor(private readonly rootDirectory = __dirname) {}

  public writeSync: runner.TextFileWriter = (filePath, content) => {
    this.files.set(path.relative(this.rootDirectory, filePath), content);
  };
}

class TestLogger {
  public text = "";

  public write = (text: string) => {
    this.text += text;
  };
}

jest.mock("../util/stopwatch", () => ({
  // Mock stopwatch to always return 42 milliseconds
  stopwatch: (fn: () => Promise<void>) => fn().then(() => 42)
}));

describe(runner.runner.name, () => {
  async function test(example: string) {
    const rootDirectory = path.join(__dirname, "examples");
    const fs = new TestTextFileWriter(rootDirectory);
    const logger = new TestLogger();
    await runner.runner(path.join(rootDirectory, example), fs.writeSync, logger);
    expect(fs.files).toMatchSnapshot();
    expect(logger.text).toMatchSnapshot();
  }

  it("project-no-config", () => test("project-no-config"));
  it("project-with-config", () => test("project-with-config"));
});
