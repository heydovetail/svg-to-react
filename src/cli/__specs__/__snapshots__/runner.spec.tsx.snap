// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runner project-no-config 1`] = `
Map {
  "project-no-config/Empty.tsx" => "// tslint:disable
//
//  This file was generated automatically by @heydovetail/svg-to-react and
//  should not be manually edited.
//
import { createSvg } from \\"@heydovetail/svg-to-react\\";
import * as React from \\"react\\";

export default createSvg((width = 24, height = 24) => {
  return <svg width={width} height={height} viewBox=\\"0 0 24 24\\" />;
});
",
}
`;

exports[`runner project-no-config 2`] = `
"Generating React components for 1 SVGs:
[33m>[39m __svgs__/empty.svg
[32mâœ” __svgs__/empty.svg[39m [90m(42ms)[39m
"
`;

exports[`runner project-with-config 1`] = `
Map {
  "project-with-config/empty.js" => "// tslint:disable
//
//  This file was generated automatically by @heydovetail/svg-to-react and
//  should not be manually edited.
//
import { createSvg } from \\"@heydovetail/svg-to-react\\";
import * as React from \\"react\\";

export default createSvg((width = 24, height = 24, color) => {
  return (
    <svg width={width} height={height} viewBox=\\"0 0 24 24\\">
      <g
        fill=\\"none\\"
        fillRule=\\"evenodd\\"
        strokeLinecap=\\"round\\"
        strokeLinejoin=\\"round\\"
        strokeWidth={2}
      >
        <circle cx={10} cy={10} r={10} stroke={color} />
        <circle cx={10} cy={10} r={10} stroke={color} />
      </g>
    </svg>
  );
});
",
}
`;

exports[`runner project-with-config 2`] = `
"Generating React components for 1 SVGs:
[33m>[39m __svgs__/empty.svg
[32mâœ” __svgs__/empty.svg[39m [90m(42ms)[39m
"
`;
